### PR TITLE
adding new CI golang tutorial

### PR DESCRIPTION
**Pull request by jimsheldon**
# Harness Developer Pull Request
Thanks for helping us make the Developer Hub better. The PR will be looked at
by the maintainers. 

## What Type of PR is This?

- [ ] Issue
- [x] Feature
- [ ] Maintenance/Chore

If tied to an Issue, list the Issue(s) here:

* *Issue(s)*

## House Keeping
Some items to keep track of. Screen shots of changes are optional but would help the maintainers review quicker. 

- [x] Tested Locally
- [ ] *Optional* Screen Shoot.

### CODE REVIEW

File: tutorials/ci-pipelines/build/ci-golang.md [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+            steps:
+              - step:
+                  type: Run
+                  identifier: installgo
+                  name: Install Go
+                  spec:
+                    shell: Sh
+                    command: |-
+                      export GOPATH=$HOME/go
+                      go install golang.org/dl/go<+matrix.goVersion>@latest
```
Comment (hemanthmantri) [inline tutorials/ci-pipelines/build/ci-golang.md:420]: Our VM image comes with a [few versions of Golang](https://github.com/wings-software/virtual-environments/blob/main/images/linux/Ubuntu2204-Readme.md#go) pre-installed. Can we mention how to switch between these already installed versions without having to install them again (unless the install is a no-op for already installed versions)? Are we no longer advocating using GH Actions for setup-<language> steps?

cc: @rustd @bradrydzewski

File: tutorials/ci-pipelines/build/ci-golang.md [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+              - step:
+                  type: Run
+                  name: Test
+                  identifier: test
+                  spec:
+                    shell: Sh
+                    command: |-
+                      export GOPATH=$HOME/go
+                      $GOPATH/bin/go<+matrix.goVersion> install github.com/jstemmer/go-junit-report/v2@latest
+                      $GOPATH/bin/go<+matrix.goVersion> test -v ./... | tee report.out
```
Comment (hemanthmantri) [inline tutorials/ci-pipelines/build/ci-golang.md:440]: Since matrix steps run in parallel, they will all write to the same `report.out`, right? Use `report_<matrix.goVersion>.out`

File: tutorials/ci-pipelines/build/ci-golang.md (hunk 1 above)
Comment (aimurphy) [reply tutorials/ci-pipelines/build/ci-golang.md:420]: For switching versions: I think that Jim included an example of switching versions - @jimsheldon ?

File: tutorials/ci-pipelines/build/ci-golang.md [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+sidebar_position: 5
+title: Build and test a Go app
+description: Use a CI pipeline to build and test a Go application.
+keywords: [Hosted Build, Continuous Integration, Hosted, CI Tutorial]
+slug: /ci-pipelines/build/golang
+---
+
+# Build and test a Go app
+
+Build and test a [Go](https://go.dev/) application using Harness CI's Linux Cloud infrastructure.
```
Comment (aimurphy) [inline tutorials/ci-pipelines/build/ci-golang.md:11]:
```suggestion
Build and test a [Go](https://go.dev/) application using a Linux platform on [Harness Cloud build infrastructure](/docs/continuous-integration/use-ci/set-up-build-infrastructure/use-harness-cloud-build-infrastructure).
```

File: tutorials/ci-pipelines/build/ci-golang.md (hunk 1 above)
Comment (bradrydzewski) [reply tutorials/ci-pipelines/build/ci-golang.md:420]: agreed, I would like to see us optimize the tutorials to take advantage of pre-installed software instead of asking our users to install software.  @hemanthmantri can you suggest the correct pipeline configuration to do a go build / test that uses the pre-installed Go version?

File: tutorials/ci-pipelines/build/ci-golang.md (hunk 2 above)
Comment (bradrydzewski) [reply tutorials/ci-pipelines/build/ci-golang.md:440]: similar to above comment, let's optimize the matrix to take advantage of the existing Go versions on the machine

**Comment by bradrydzewski** [comment]
The tutorial starts out by installing Go, but we should have Go pre-installed on the machine.  @hemanthmantri can you provide an example yaml that uses the pre-installed version of Go?  I would like to optimize the tutorials to use the pre-installed software where possible.

**Comment by rustd** [comment]
+1 to Brad. Thank you for writing this Jim. This is simple and concise. Few suggestions to simplify it further.

- Remove the step to install Go since it's already installed on the VM
- Remove "Install a specific version of Go" & "Install multiple versions of Go" samples for all sub-sections (Install, Cache, Run tests)
- If you want you can add an optional section at the end to change the go version on the VM.

File: tutorials/ci-pipelines/build/ci-golang.md (hunk 2 above)
Comment (jimsheldon) [reply tutorials/ci-pipelines/build/ci-golang.md]: Thanks for catching this, I pushed a fix.

File: tutorials/ci-pipelines/build/ci-golang.md (hunk 1 above)
Comment (jimsheldon) [reply tutorials/ci-pipelines/build/ci-golang.md:421]: @hemanthmantri there is a link to documentation under "Setup Go" that describes the OS and tools installed.

The versions of Go are installed under `/opt/hostedtoolcache/go`, I need to work with someone who can help me recommend the best method if switching between these versions.

File: tutorials/ci-pipelines/build/ci-golang.md [hunk 4]
```diff
@@ -0,0 +1,648 @@
...
+</Tabs>
+```
+
+## Visualize test results
+
+Test results can be [viewed](/docs/continuous-integration/use-ci/set-up-test-intelligence/viewing-tests/) in the **Tests** tab in your pipeline execution. Test results must be in JUnit XML format.
+
+[go-junit-report](https://github.com/jstemmer/go-junit-report) can be used to output compatible JUnit XML reports.
+
+Modify your test step to generate the JUnit XML, and add the reports path.
```
Comment (aimurphy) [inline tutorials/ci-pipelines/build/ci-golang.md:163]:
```suggestion
For your pipeline to produce test reports, you need to modify the **Run** step that runs your tests. Make sure the `command` generates JUnit XML reports and add the `reports` specification.
```

I am concerned that calling it "tests" might make people think there's a "tests" step or get confused with the "Run tests" step. Feel free to reword this as you like.

Comment (aimurphy) [inline tutorials/ci-pipelines/build/ci-golang.md:163]:
```suggestion
For your pipeline to publish test reports to the **Tests** tab, you need to modify the **Run** step that runs your tests. Make sure the `command` generates JUnit XML reports and add the `reports` specification.
```

I am concerned that calling it "tests" might make people think there's a "tests" step or get confused with the "Run tests" step. Feel free to reword this as you like.

File: tutorials/ci-pipelines/build/ci-golang.md [hunk 5]
```diff
@@ -0,0 +1,648 @@
...
+```
+
+## Specify version
+
+```mdx-code-block
+<Tabs>
+<TabItem value="Harness Cloud">
+```
+
+Hosted Cloud runners come with Go pre-installed, see [Platforms and image specifications](/docs/continuous-integration/use-ci/set-up-build-infrastructure/use-harness-cloud-build-infrastructure#platforms-and-image-specifications) for all available tools and versions.
```
Comment (aimurphy) [inline tutorials/ci-pipelines/build/ci-golang.md:225]:
```suggestion
Go is pre-installed on Hosted Cloud runners. For details about all available tools and versions, go to [Platforms and image specifications](/docs/continuous-integration/use-ci/set-up-build-infrastructure/use-harness-cloud-build-infrastructure#platforms-and-image-specifications).
```

File: tutorials/ci-pipelines/build/ci-golang.md [hunk 6]
```diff
@@ -0,0 +1,648 @@
...
+                  type: Run
+                  identifier: installgo
+                  name: Install Go
+                  spec:
+                    shell: Sh
+                    command: |-
+                      export GOPATH=$HOME/go
+                      go install golang.org/dl/go<+matrix.goVersion>@latest
+                      $GOPATH/bin/go<+matrix.goVersion> download
+```
```
Comment (aimurphy) [inline tutorials/ci-pipelines/build/ci-golang.md:275]:
~~~suggestion
1. Add the [matrix looping strategy](/docs/platform/pipelines/looping-strategies-matrix-repeat-and-parallelism/) configuration to your stage.

   ```yaml
           strategy:
             matrix:
               # matrix strategy with Go versions 1.19 and 1.20
               goVersion:
                 - "1.19"
                 - "1.20"
   ```

2. Reference the matrix variable in your steps.

   ```yaml
                 - step:
                     type: Run
                     identifier: installgo
                     name: Install Go
                     spec:
                       shell: Sh
                       command: |-
                         export GOPATH=$HOME/go
                         go install golang.org/dl/go<+matrix.goVersion>@latest
                         $GOPATH/bin/go<+matrix.goVersion> download
   ```
~~~

Since there are two steps separated by a code block, I think the numbering might help the second step stick out.

File: tutorials/ci-pipelines/build/ci-golang.md [hunk 7]
```diff
@@ -0,0 +1,648 @@
...
+
+</details>
+
+
+
+<details>
+<summary>Build using multiple versions of Go</summary>
+
+Add the [matrix looping strategy](/docs/platform/pipelines/looping-strategies-matrix-repeat-and-parallelism/) configuration to your stage.
```
Comment (aimurphy) [inline tutorials/ci-pipelines/build/ci-golang.md:310]: same suggestion here to use numbering

File: tutorials/ci-pipelines/build/ci-golang.md [hunk 8]
```diff
@@ -0,0 +1,648 @@
...
+<Tabs>
+<TabItem value="Harness Cloud">
+```
+
+Replace the bracketed values with corresponding values for your [code repo connector](/docs/continuous-integration/use-ci/codebase-configuration/create-and-configure-a-codebase/#code-repo-connectors) and repository name.
+
+Depending on your project and organization, you may also need to replace `projectIdentifier` and `orgIdentifier`.
+
+<details>
+<summary>Use a specific Go version</summary>
```
Comment (aimurphy) [inline tutorials/ci-pipelines/build/ci-golang.md:356]:
```suggestion
<summary>Pipeline with one specific Go version</summary>
```

File: tutorials/ci-pipelines/build/ci-golang.md [hunk 9]
```diff
@@ -0,0 +1,648 @@
...
+      codebase:
+        connectorRef: [your-code-repo-connector-ID] # replace with your connector ID
+        repoName: [your-repository-name] # replace with your repository name
+        build: <+input>
+```
+
+</details>
+
+<details>
+<summary>Use multiple Go versions</summary>
```
Comment (aimurphy) [inline tutorials/ci-pipelines/build/ci-golang.md:427]:
```suggestion
<summary>Pipeline with multiple Go versions</summary>
```

File: tutorials/ci-pipelines/build/ci-golang.md [hunk 10]
```diff
@@ -0,0 +1,648 @@
...
+</details>
+
+```mdx-code-block
+</TabItem>
+</Tabs>
+```
+
+## Full pipeline examples
+
+Full pipeline examples based on the steps above.
```
Comment (aimurphy) [inline tutorials/ci-pipelines/build/ci-golang.md:344]: I don't see caching in these example so maybe clarify like:

These YAML examples define a single-stage pipeline with steps to install Go, build a Go app, and run tests.

File: tutorials/ci-pipelines/build/ci-golang.md [hunk 11]
```diff
@@ -0,0 +1,648 @@
...
+
+<TabItem value="Self-hosted">
+```
+
+Replace the bracketed values with corresponding values for your [code repo connector](/docs/continuous-integration/use-ci/codebase-configuration/create-and-configure-a-codebase/#code-repo-connectors), [kubernetes cluster connector](/docs/platform/Connectors/Cloud-providers/add-a-kubernetes-cluster-connector), kubernetes namespace, and repository name.
+
+Depending on your project and organization, you may also need to replace `projectIdentifier` and `orgIdentifier`.
+
+<details>
+<summary>Use a specific Go version</summary>
```
Comment (aimurphy) [inline tutorials/ci-pipelines/build/ci-golang.md:513]:
```suggestion
<summary>Pipeline with one specific Go version</summary>
```

File: tutorials/ci-pipelines/build/ci-golang.md [hunk 12]
```diff
@@ -0,0 +1,648 @@
...
+      codebase:
+        connectorRef: [your-code-repo-connector-ID] # replace with your connector ID
+        repoName: [your-repository-name] # replace with your repository name
+        build: <+input>
+```
+
+</details>
+
+<details>
+<summary>Use multiple Go versions</summary>
```
Comment (aimurphy) [inline tutorials/ci-pipelines/build/ci-golang.md:574]:
```suggestion
<summary>Pipeline with multiple Go versions</summary>
```

File: tutorials/ci-pipelines/build/ci-golang.md [hunk 13]
```diff
@@ -0,0 +1,648 @@
...
+```mdx-code-block
+</TabItem>
+</Tabs>
+```
+
+## Next steps
+
+Create [triggers](https://developer.harness.io/docs/category/triggers) to execute your pipeline.
+
+Learn how to [build and upload artifacts](/docs/category/build-and-upload-artifacts), or how to [build and push an image to a Docker registry](/docs/continuous-integration/use-ci/build-and-upload-artifacts/build-and-push-to-docker-hub-step-settings/).
```
Comment (aimurphy) [inline tutorials/ci-pipelines/build/ci-golang.md:648]:
```suggestion
Now that you have created a pipeline that builds and tests a Go app, you could:

* Create [triggers](https://developer.harness.io/docs/category/triggers) to automatically run your pipeline.
* Add steps to [build and upload artifacts](/docs/category/build-and-upload-artifacts).
* Add a step to [build and push an image to a Docker registry](/docs/continuous-integration/use-ci/build-and-upload-artifacts/build-and-push-to-docker-hub-step-settings/).
```

File: tutorials/ci-pipelines/build/ci-golang.md (hunk 4 above)
Comment (jimsheldon) [reply tutorials/ci-pipelines/build/ci-golang.md:163]: This seems clear to me, thanks

File: tutorials/ci-pipelines/build/ci-golang.md (hunk 6 above)
Comment (jimsheldon) [reply tutorials/ci-pipelines/build/ci-golang.md:275]: Makes sense

File: tutorials/ci-pipelines/build/ci-golang.md (hunk 10 above)
Comment (jimsheldon) [reply tutorials/ci-pipelines/build/ci-golang.md:344]: Caching is in the cloud examples, but not self-hosted

I can add some details around what is featured in these pipelines

File: tutorials/ci-pipelines/build/ci-golang.md [hunk 14]
```diff
@@ -0,0 +1,648 @@
...
+                  type: Run
+                  identifier: test
+                  name: Test
+                  spec:
+                    connectorRef: [your-docker-connector-ID] # replace with your connector ID
+                    image: golang
+                    command: |-
+                      go test -v ./...
+```
+
```
[the remark this answers is not in the document]
Comment (jimsheldon) [reply tutorials/ci-pipelines/build/ci-golang.md:77]: I think we will hold off on this for now, just trying to keep this minimal if possible

**Review by aimurphy — APPROVED** [review]
lgtm if you are ready to merge it

**Review by aimurphy — APPROVED** [review]
